### PR TITLE
feat: Allow individual gatherContext for each additional output

### PR DESCRIPTION
**Pull request by Funatiq**
Summary: 
* Add support for a gatherContext flag for each additional output in ExecutorConfig.
* Enhance copyAdditionalOutputs function to work with AdditionalModelOutput.
* Add validation for final indices of additional output tensors.
* Update unit tests to verify the changes.
  * Cover the new gatherContext parameter.
  * Add lastTokenIds input tensor to test engines.
  * Make logits output dependent on gatherContextLogits parameter.

### CODE REVIEW

File: cpp/tensorrt_llm/batch_manager/runtimeBuffers.cpp
Comment (jdebache) [inline cpp/tensorrt_llm/batch_manager/runtimeBuffers.cpp:1]: Unused headers: `"tensorrt_llm/runtime/runtimeKernels.h"` and `<valarray>`

File: cpp/tensorrt_llm/pybind/bindings.cpp [hunk 1]
```diff
@@ -523,8 +523,7 @@ PYBIND11_MODULE(TRTLLM_PYBIND_MODULE, m)
         .def_readwrite("gpu_weights_percent", &tb::TrtGptModelOptionalParams::gpuWeightsPercent)
         .def_readwrite("max_beam_width", &tb::TrtGptModelOptionalParams::maxBeamWidth)
         .def_readwrite("scheduler_config", &tb::TrtGptModelOptionalParams::schedulerConfig)
-        .def(py::pickle(gptModelParamsGetState, gptModelParamsSetState))
-        .def("__eq__", &tb::TrtGptModelOptionalParams::operator==);
```
Comment (jdebache) [inline cpp/tensorrt_llm/pybind/bindings.cpp:527]: Wasn't there a place where we actually use this equality operator? It sounds vaguely familiar.

Comment (Funatiq) [reply cpp/tensorrt_llm/pybind/bindings.cpp:527]: It was used in the pickle test, but I got too annoyed to update it every time, so I changed the test too. IMO `TrtGptModelOptionalParams` shouldn't be exposed at all.

**Review by jdebache — APPROVED** [review]
